### PR TITLE
Progress towards possibility of variable-sized segments

### PR DESCRIPTION
**Pull request by jaschiu**
Idea is that to support variable-sized segments, one big problem is to analyze the map for the space available in front of the player. This PR will try to solve it with a pure function that will calculate possible dimensions for potential segments based on given data. This means, the JavaScript will also need to be modified to provide more data. I have only wrote detailed documentation of this function as of now, more commits of actual code will come soon.

### CODE REVIEW

**Comment by pcrain** [comment]
Looks good so far! 😃 Will look over again and hopefully merge once you've finished the implementation.

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,163 @@
-*/__pycache__/*
-*.pytest_cache/*
+<<<<<<< HEAD
```
Comment (wadefagen) [inline .gitignore]: This has git merge markers in the file itself -- please fix these

**Review by wadefagen — CHANGES_REQUESTED** [review]
Noticed some merge markers in a quick glance at this code

**Comment by jaschiu** [comment]
Main function is finished. I can try to analyze and modify the middleware/frontend a bit to send more data and modify which MGs are chosen, but may take a few days without help from other people on this issue.

Edit: I also fixed my .gitignore having a merge artificat but I don't know how to get the Changes Requested notification to go away

**Comment by jaschiu** [comment]
Modifications needed for middleware:

- requests to middleware's `/generateSegment` should include the direction the player intended to move in.
- the data structure storing the global segment stores each individual coordinate as keys. I would have to convert it to a collection of square boundaries, wasting computational resources, so why not store the square boundaries *as* the keys?
- MGs will have to incorporate more data, such as the offsets of all of their exits on each side, hopefully it's not too much of a drastic change
- may have to make the random MG chooser more complex?

**Comment by wadefagen** [comment]
> Edit: I also fixed my .gitignore having a merge artificat but I don't know how to get the Changes Requested notification to go away

On the reviewer area, you should be able to press the "refresh" button next to the name of the person who requested the change:
image

**Comment by jaschiu** [comment]
> > Edit: I also fixed my .gitignore having a merge artificat but I don't know how to get the Changes Requested notification to go away
> 
> On the reviewer area, you should be able to press the "refresh" button next to the name of the person who requested the change: image

Done.

I'm going to try to spend some time figuring out how to modify some stuff in the middleware right now.

**Comment by jaschiu** [comment]
I changed `app.py` `generateSegment` to make use of the function, but a lot of the middleware needs to have a different structure. I may not be able to finish this by today, but other people can help out.

**Comment by pcrain** [comment]
@jasyip I think some of the other students were able to incorporate the ideas and functionality you started into the shared middleware. I don't know if your commits will show up in the history but we'll make sure you get credit either way. Thank you very much for all of your hard work on the variable-sized segments! :smile:

**Pull request closed without merging**